### PR TITLE
chore(warehouse): use fastUUID with google UUID generation

### PR DESCRIPTION
**Pull request by achettyiitr**
# Description

**Use google's uuid, as gofrs can cause [performance issues](https://github.com/rudderlabs/rudder-server/pull/1606)** 

**Also, we have introduced one more utility called RandHex which generated a random string of 32 characters from UUID.** 
(The benefit here is that at some places where after getting the UUID string of 36 characters we used to replace `-`. Since we are doing extra computation of preparing the UUID string and replacing the `-`.  The same functionality is being provided by RandHex.)

## Notion Ticket

https://www.notion.so/rudderstacks/Fasf-UUID-4683365544654681aa04f69f790e05ce

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: warehouse/validations/helper.go [hunk 1]
```diff
@@ -59,7 +58,7 @@ func parseOptions(req json.RawMessage, v interface{}) error {
 }
 
 func randomString() string {
-	return strings.ReplaceAll(uuid.Must(uuid.NewV4()).String(), "-", "")
+	return strings.ReplaceAll(misc.FastUUID().String(), "-", "")
```
Comment (lvrach) [inline warehouse/validations/helper.go:61]: There is probably a faster way to do this by converting the UUID bytes directly to runes. If this is used in a performance-critical path, it is probably worth it ( a benchmark will be advised)

Comment (lvrach) [reply warehouse/validations/helper.go:61]: We can use our own, modified version of encodeHex(dst []byte, uuid UUID) https://github.com/google/uuid/blob/44b5fee7c49cf3bcdf723f106b36d56ef13ccc88/uuid.go#L200

A single call to [hex.Encode](https://pkg.go.dev/encoding/hex#Encode) should be enough.

Comment (lvrach) [reply warehouse/validations/helper.go:61]:
```suggestion
	var buf [32]byte
	 hex.Encode(buf[:], misc.FastUUID())
	return string(buf[:])
```

Comment (achettyiitr) [reply warehouse/validations/helper.go:61]: Make sense. Since we are ignoring `-`. It's better to use this.

Comment (achettyiitr) [reply warehouse/validations/helper.go]: Done.

File: warehouse/utils/utils.go [hunk 2]
```diff
@@ -1074,8 +1074,15 @@ func StagingTablePrefix(provider string) string {
...
 
 func StagingTableName(provider, tableName string, tableNameLimit int) string {
-	randomNess := strings.ReplaceAll(uuid.Must(uuid.NewV4()).String(), "-", "")
+	randomNess := RandStringWithoutQuotes()
 	prefix := StagingTablePrefix(provider)
 	stagingTableName := fmt.Sprintf(`%s%s_%s`, prefix, tableName, randomNess)
 	return misc.TruncateStr(stagingTableName, tableNameLimit)
 }
+
+func RandStringWithoutQuotes() string {
```
[the remark this answers is not in the document]
Comment (achettyiitr) [reply warehouse/utils/utils.go:1083]: Done.

[the remark this answers is not in the document]
Comment (achettyiitr) [reply warehouse/utils/utils.go:1083]: Done.

File: warehouse/bigquery/bigquery.go [hunk 3]
```diff
@@ -276,7 +276,7 @@ func (bq *HandleT) DeleteBy(tableNames []string, params warehouseutils.DeleteByP
 }
 
 func partitionedTable(tableName, partitionDate string) string {
-	return fmt.Sprintf(`%s$%v`, tableName, strings.ReplaceAll(partitionDate, "-", ""))
+	return fmt.Sprintf(`%s$%v`, tableName, warehouseutils.RandStringWithoutQuotes())
```
[the remark this answers is not in the document]
Comment (achettyiitr) [reply warehouse/bigquery/bigquery.go:279]: Thanks, @lvrach for pointing it out. It happened by mistake.

**Review by lvrach — APPROVED** [review]
Thanks for addressing my comments

File: warehouse/utils/utils.go [hunk 4]
```diff
@@ -1075,8 +1075,16 @@ func StagingTablePrefix(provider string) string {
...
 func StagingTableName(provider, tableName string, tableNameLimit int) string {
-	randomNess := strings.ReplaceAll(uuid.Must(uuid.NewV4()).String(), "-", "")
+	randomNess := RandHex()
 	prefix := StagingTablePrefix(provider)
 	stagingTableName := fmt.Sprintf(`%s%s_%s`, prefix, tableName, randomNess)
 	return misc.TruncateStr(stagingTableName, tableNameLimit)
 }
+
+// RandHex returns a random hex string of length 32
+func RandHex() string {
```
Comment (abhimanyubabbar) [inline warehouse/utils/utils.go:1085]: Is there any improvement in this function in terms of speed or uniqueness over the other function ? Just curious to understand the motivation behind it.

**Review by abhimanyubabbar — APPROVED** [review]
Overall the changes looks good. 👍🏼 LGTM
I have just one comment on the motivation behind creation of RandomHex() function.

File: warehouse/utils/utils.go (hunk 4 above)
Comment (achettyiitr) [reply warehouse/utils/utils.go:1085]: Yes. Actually, UUID.string() is processing for 36 bytes. We are ignoring the `-` present by replacing it using `strings.ReplaceAll(uuid.Must(uuid.NewV4()).String(), "-", "")`. We can just probably call `hex.Encode` which can just populate the 32 bytes.

```
const Size = 16
type UUID [Size]byte
```

```
func (u UUID) String() string {
	buf := make([]byte, 36)

	hex.Encode(buf[0:8], u[0:4])
	buf[8] = '-'
	hex.Encode(buf[9:13], u[4:6])
	buf[13] = '-'
	hex.Encode(buf[14:18], u[6:8])
	buf[18] = '-'
	hex.Encode(buf[19:23], u[8:10])
	buf[23] = '-'
	hex.Encode(buf[24:], u[10:])

	return string(buf)
}
```

**Pull request merged**